app/views/successes/index.html.erb

```diff
@@ -12,6 +12,8 @@
         <% @google_map_metadata.each do |location| %>
             <li data-lat="<%= location[:latitude] %>"
                 data-lon="<%= location[:longitude] %>"
+                data-name="<%= location[:dog_name] %>"
```

Comment:
Yes, it's done :muscle:  But I was thinking about updating the integration test, that's why it was [WIP]. However, since we are applying random values to `lon` and `lat` it's not very easy to test properly. Besides this, It's all done and working. You can merge this branch if you want to and we can update the test later? @kasugaijin

Reply:
Ah yes, the way the `DataBuilder::GoogleMap` class is constructed right now is it's receiving an active record association of the adoptions, with the dog and locations data already eager loaded. So, we have all the data we need in the `adoptions` collection passed to `DataBuilder::GoogleMap`. However, within the service object we are makin another database call to retrieve the location data for the `@location` instance variable, then using that to apply deviations. This is making it way harder to test, and is less efficient. We already have the data so there is no need to query the database again. My bad for not realizing this earlier...it was only when I started thinking about testing this service object. 

If we were to use the collection data passed to the service object, we could check that each adoption has a lat and a lon, and if so, apply a random deviation to it without any need for database queries. This will also enable us to test the implementation of this service object much easier. We could set up a unit test by feeding the class plug data (an array with one object with lat, lon, name, breed) and then call the `generate_map_coordinates` method and assert the output has different lat and lon from the plug data. Better still, we could stub the `rand` method and pass it a known value and assert the lat and lon have changed by that value.

What are your thoughts on this?

Reply:
> Ah yes, the way the `DataBuilder::GoogleMap` class is constructed right now is it's receiving an active record association of the adoptions, with the dog and locations data already eager loaded. So, we have all the data we need in the `adoptions` collection passed to `DataBuilder::GoogleMap`. However, within the service object we are makin another database call to retrieve the location data for the `@location` instance variable, then using that to apply deviations. This is making it way harder to test, and is less efficient. We already have the data so there is no need to query the database again. My bad for not realizing this earlier...it was only when I started thinking about testing this service object. 
> 
> If we were to use the collection data passed to the service object, we could check that each adoption has a lat and a lon, and if so, apply a random deviation to it without any need for database queries. This will also enable us to test the implementation of this service object much easier. We could set up a unit test by feeding the class plug data (an array with one object with lat, lon, name, breed) and then call the `generate_map_coordinates` method and assert the output has different lat and lon from the plug data. Better still, we could stub the `rand` method and pass it a known value and assert the lat and lon have changed by that value.
> 
> What are your thoughts on this?

I'm on mobile right now, and I'll be away from my pc for a few days, but how exactly are we making another database call to retrieve location data? Are we not utilizing the already loaded data? I remember that we are using eager load to get what is needed beforehand.

Instead of checking if each adoption has lat and lon, it's better to use validation upon creation. If there's already validations, I see no reason to check this in service object.

`generate_map_coordinates` method is a private class method. We can't call it directly and feed it with an array. We can use fixtures, and if I am not mistaken, I remember that the test database is generated from fixtures. So we don't need to feed. We just need to set the test up in a way that simulates our real-world usage. I think the current way that the service object test is set up is okay (Previous PR). We can modify it still.

As you mentioned, unit testing would be much better with assert_different. Maybe we don't need to test if html has an element with data attributes. In the end, it's basic rails, so there is no reason to test rails if we have validations.

What do you think?

Reply:
Yes you are right, there is not another database call...it is using the eager loaded data. I don't know why I thought differently there. Derp.

The benefit of checking for lat/lon on each adoption is because a lat/lon is not a guarantee on each location. The way the lat/lon is generated is using a gem called geocoder, which takes an address (e.g., streen, city, country) and pings the google API to return a lat/lon and then persist that data. For several reasons, this might fail, whether it's a network error or a nil result from google because it cannot parse the address data, it may not get generated. For example, in some rare cases, someone might just enter a gibberish address to pass the validation that requires a value...e.g., 123 ABC st, Nowhere, Space or something to that effect. This will fail on the geocoder coordinate generation. Therefore, we cannot rely on validation for that. There is probably a more sophisticated way to to this, i.e. take an address, and then return a validation error if the coordinates don't get generated, but currently that is not the case.

Looking at the fixtures, there is one adoption, linked to adopter_account_one, which has lat/lon in its associated location fixture. Without thinking about this in detail, it seems plausible to access the /successes route and assert that the location data in the html is different from that in the fixture? That would test the entire implementation, as it would be used in the real world. Or am I missing something here?

Reply:
It all makes sense to me. I will take a deeper look at when I get back to my computer ✔️

Reply:
Sounds good!

Reply:
I added your suggestion, I think you can merge it.